src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -59,6 +59,17 @@ void GcsNodeManager::HandleRegisterNode(rpc::RegisterNodeRequest request,
     RAY_LOG(INFO) << "Finished registering node info, node id = " << node_id
                   << ", address = " << request.node_info().node_manager_address()
                   << ", node name = " << request.node_info().node_name();
+    if (request.node_info().is_head_node()) {
+      // mark the old head node as dead
```

Comment:
```suggestion
      // mark the old head node as dead if exists
```

Reply:
Also comment when this happens 

```
1. should never happen when HA is not used
2. happens when a new GCS is started
```

Reply:
I think the time complexity will not change to O(N^2) since we only iterate through all node when we add head node which only happens once per gcs server creation (since if a head node dead, then gcs_server also dead?), so same frequency as num calls of `Initialize` which also iterate through all alive nodes.

Reply:
ah yeah. I missed that part. sgtm then. 

For finding a head node, is it possible to find "all" nodes with head = true? I think it is safer to not assume there's only 1 additional head node (although in reality it could be true)